core/endorser/endorser.go

```diff
@@ -248,6 +253,97 @@ func (e *Endorser) SimulateProposal(txParams *ccprovider.TransactionParams, chai
+	// early, so this is invoked multiple times, but that is how the code worked before
+	// this change, so, should be safe.  Long term, let's move the Done up to the create.
+	defer txParams.TXSimulator.Done()
+
+	simResult, err := txParams.TXSimulator.GetTxSimulationResults()
+	if err != nil {
+		e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+		return nil, nil, nil, nil, err
+	}
+
+	if simResult.PvtSimulationResults != nil {
+		if chaincodeName == "lscc" {
+			// TODO: remove once we can store collection configuration outside of LSCC
+			e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+			return nil, nil, nil, nil, errors.New("Private data is forbidden to be used in instantiate")
+		}
+		pvtDataWithConfig, err := AssemblePvtRWSet(txParams.ChannelID, simResult.PvtSimulationResults, txParams.TXSimulator, e.Support.GetDeployedCCInfoProvider())
+		// To read collection config need to read collection updates before
+		// releasing the lock, hence txParams.TXSimulator.Done()  moved down here
+		txParams.TXSimulator.Done()
+
+		if err != nil {
+			e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+			return nil, nil, nil, nil, errors.WithMessage(err, "failed to obtain collections config")
+		}
+		endorsedAt, err := e.Support.GetLedgerHeight(txParams.ChannelID)
+		if err != nil {
+			e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+			return nil, nil, nil, nil, errors.WithMessage(err, fmt.Sprintf("failed to obtain ledger height for channel '%s'", txParams.ChannelID))
+		}
+		// Add ledger height at which transaction was endorsed,
+		// `endorsedAt` is obtained from the block storage and at times this could be 'endorsement Height + 1'.
+		// However, since we use this height only to select the configuration (3rd parameter in distributePrivateData) and
+		// manage transient store purge for orphaned private writesets (4th parameter in distributePrivateData), this works for now.
+		// Ideally, ledger should add support in the simulator as a first class function `GetHeight()`.
+		pvtDataWithConfig.EndorsedAt = endorsedAt
+		if err := e.PrivateDataDistributor.DistributePrivateData(txParams.ChannelID, txParams.TxID, pvtDataWithConfig, endorsedAt); err != nil {
+			e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+			return nil, nil, nil, nil, err
+		}
+	}
+
+	// Yacov: This is where the writes that appear on the blockchain are assembled,
+	// right before they are passed to EndorseWithPlugin (line 451).
+
+	// GAL: pass by reference on simResults and add hashes
+	pubSimResBytes, pis, err := simResult.GetPubSimulationBytesGDPR(helperGDPR)
+	if err != nil {
+		e.Metrics.SimulationFailure.With(meterLabels...).Add(1)
+		return nil, nil, nil, nil, err
+	}
+
+	return res, pubSimResBytes, ccevent, pis, nil
+}
+
+func helperGDPR(nsrws *rwset.NsReadWriteSet) (*rwset.NsReadWriteSet, [][]byte) {
+	pis := make([][]byte, 100)
+	rwset := &rwsetutil.TxRwSet{}
+	rwset.FromProtoBytes(nsrws.Rwset)
```

Comment:
what if `rwset.FromProtoBytes(nsrws.Rwset)` fails? We need to check the error status

Reply:
I think this should be `KVRWSet`